course_UvA-DL/vision-transformer/Vision_Transformer.py

```diff
@@ -0,0 +1,526 @@
+    if "/" in file_name.split("/", 1)[1]:
+        os.makedirs(file_path.rsplit("/", 1)[0], exist_ok=True)
+    if not os.path.isfile(file_path):
+        file_url = base_url + file_name
+        print("Downloading %s..." % file_url)
+        try:
+            urllib.request.urlretrieve(file_url, file_path)
+        except HTTPError as e:
+            print(
+                "Something went wrong. Please try to download the file from the GDrive folder, or contact the author with the full output including the following error:\n",
+                e
+            )
+
+# %% [markdown]
+# We load the CIFAR10 dataset below.
+# We use the same setup of the datasets and data augmentations as for the CNNs in Tutorial 5 to keep a fair comparison.
+# The constants in the `transforms.Normalize` correspond to the values
+# that scale and shift the data to a zero mean and standard deviation of
+# one.
+
+# %%
+test_transform = transforms.Compose([
+    transforms.ToTensor(),
+    transforms.Normalize([0.49139968, 0.48215841, 0.44653091], [0.24703223, 0.24348513, 0.26158784])
+])
+# For training, we add some augmentation. Networks are too powerful and would overfit.
+train_transform = transforms.Compose([
+    transforms.RandomHorizontalFlip(),
+    transforms.RandomResizedCrop((32, 32), scale=(0.8, 1.0), ratio=(0.9, 1.1)),
+    transforms.ToTensor(),
+    transforms.Normalize([0.49139968, 0.48215841, 0.44653091], [0.24703223, 0.24348513, 0.26158784])
+])
+# Loading the training dataset. We need to split it into a training and validation part
+# We need to do a little trick because the validation set should not use the augmentation.
+train_dataset = CIFAR10(root=DATASET_PATH, train=True, transform=train_transform, download=True)
+val_dataset = CIFAR10(root=DATASET_PATH, train=True, transform=test_transform, download=True)
+pl.seed_everything(42)
+train_set, _ = torch.utils.data.random_split(train_dataset, [45000, 5000])
+pl.seed_everything(42)
+_, val_set = torch.utils.data.random_split(val_dataset, [45000, 5000])
+
+# Loading the test set
+test_set = CIFAR10(root=DATASET_PATH, train=False, transform=test_transform, download=True)
+
+# We define a set of data loaders that we can use for various purposes later.
+train_loader = data.DataLoader(
+    train_set, batch_size=128, shuffle=True, drop_last=True, pin_memory=True, num_workers=4
+)
+val_loader = data.DataLoader(val_set, batch_size=128, shuffle=False, drop_last=False, num_workers=4)
+test_loader = data.DataLoader(test_set, batch_size=128, shuffle=False, drop_last=False, num_workers=4)
+
+# Visualize some examples
+NUM_IMAGES = 4
+CIFAR_images = torch.stack([val_set[idx][0] for idx in range(NUM_IMAGES)], dim=0)
+img_grid = torchvision.utils.make_grid(CIFAR_images, nrow=4, normalize=True, pad_value=0.9)
+img_grid = img_grid.permute(1, 2, 0)
+
+plt.figure(figsize=(8, 8))
+plt.title("Image examples of the CIFAR10 dataset")
+plt.imshow(img_grid)
+plt.axis('off')
+plt.show()
+plt.close()
```

Comment:
not sure if this makes sense in the context of a notebook (for the script it may freeze on show, right?)

Reply:
That is actually a good point. The script could freeze because of this, but I did not experience it in our tests so far here.